File: python/src/ya_obs/_xml.py
from __future__ import annotations
import xml.etree.ElementTree as ET
from datetime import datetime, timezone


_NS = "http://obs.myhwclouds.com/doc/2015-06-30/"


def _tag(name: str) -> str:
    return f"{{{_NS}}}{name}"


def _text(el: ET.Element, tag: str, default: str = "") -> str:
    child = el.find(_tag(tag))
    return child.text or default if child is not None else default


def parse_list_objects(xml: str) -> dict:
    root = ET.fromstring(xml)
    objects = []
    for content in root.findall(_tag("Contents")):
        last_modified_str = _text(content, "LastModified")
        last_modified = datetime.fromisoformat(
            last_modified_str.replace("Z", "+00:00")
        )
        objects.append({
            "key": _text(content, "Key"),
            "etag": _text(content, "ETag"),
            "size": int(_text(content, "Size", "0")),
            "last_modified": last_modified.isoformat(),
        })
    next_marker_el = root.find(_tag("NextMarker"))
    return {
        "name": _text(root, "Name"),
        "is_truncated": _text(root, "IsTruncated").lower() == "true",
        "next_marker": next_marker_el.text if next_marker_el is not None and next_marker_el.text else None,
        "objects": objects,
    }


def parse_initiate_multipart(xml: str) -> dict:
    root = ET.fromstring(xml)
    return {
        "bucket": _text(root, "Bucket"),
        "key": _text(root, "Key"),
        "upload_id": _text(root, "UploadId"),
    }


def parse_error_response(xml: str) -> dict:
    root = ET.fromstring(xml)
    def t(name: str) -> str:
        el = root.find(name)
        return el.text or "" if el is not None else ""
    return {
        "code": t("Code"),
        "message": t("Message"),
        "request_id": t("RequestId"),
        "host_id": t("HostId"),
    }


def serialize_complete_multipart(parts: list[tuple[int, str]]) -> str:
    root = ET.Element("CompleteMultipartUpload")
    for part_number, etag in parts:
        part_el = ET.SubElement(root, "Part")
        pn_el = ET.SubElement(part_el, "PartNumber")
        pn_el.text = str(part_number)
        etag_el = ET.SubElement(part_el, "ETag")
        etag_el.text = etag
    return ET.tostring(root, encoding="unicode")

File: rust/ya-obs/src/xml.rs
use quick_xml::de::from_str;
use serde::Deserialize;

use crate::models::{
    ErrorResponse, InitiateMultipartResult, ListBucketResult, ListedBucket, ListedObject,
};

#[derive(Debug, Deserialize)]
struct RawContents {
    #[serde(rename = "Key")]
    key: String,
    #[serde(rename = "LastModified")]
    last_modified: String,
    #[serde(rename = "ETag")]
    etag: String,
    #[serde(rename = "Size")]
    size: u64,
}

#[derive(Debug, Deserialize)]
struct RawListBucketResult {
    #[serde(rename = "Name")]
    name: String,
    #[serde(rename = "Prefix", default)]
    prefix: Option<String>,
    #[serde(rename = "MaxKeys", default)]
    max_keys: u32,
    #[serde(rename = "IsTruncated")]
    is_truncated: bool,
    #[serde(rename = "NextMarker", default)]
    next_marker: Option<String>,
    #[serde(rename = "Contents", default)]
    contents: Vec<RawContents>,
}

/// Normalize OBS `LastModified` (`2024-01-15T10:30:00.000Z`) into the
/// `2024-01-15T10:30:00+00:00` form the test vectors expect.
fn normalize_last_modified(raw: &str) -> String {
    let no_frac = raw.split('.').next().unwrap_or(raw);
    if raw.ends_with('Z') {
        format!("{no_frac}+00:00")
    } else {
        no_frac.to_string()
    }
}

#[derive(Debug, Deserialize)]
struct RawBucketEntry {
    #[serde(rename = "Name")]
    name: String,
    #[serde(rename = "CreationDate")]
    creation_date: String,
}

#[derive(Debug, Deserialize)]
struct RawBuckets {
    #[serde(rename = "Bucket", default)]
    bucket: Vec<RawBucketEntry>,
}

#[derive(Debug, Deserialize)]
struct RawListAllMyBucketsResult {
    #[serde(rename = "Buckets", default)]
    buckets: Option<RawBuckets>,
}

pub fn parse_list_all_my_buckets(xml: &str) -> Result<Vec<ListedBucket>, quick_xml::DeError> {
    let raw: RawListAllMyBucketsResult = from_str(xml)?;
    Ok(raw
        .buckets
        .map(|b| b.bucket)
        .unwrap_or_default()
        .into_iter()
        .map(|b| ListedBucket {
            name: b.name,
            creation_date: normalize_last_modified(&b.creation_date),
        })
        .collect())
}

pub fn parse_list_bucket_result(xml: &str) -> Result<ListBucketResult, quick_xml::DeError> {
    let raw: RawListBucketResult = from_str(xml)?;

    let objects = raw
        .contents
        .into_iter()
        .map(|c| ListedObject {
            key: c.key,
            etag: c.etag,
            size: c.size,
            last_modified: normalize_last_modified(&c.last_modified),
        })
        .collect();

    Ok(ListBucketResult {
        name: raw.name,
        prefix: raw.prefix.filter(|p| !p.is_empty()),
        max_keys: raw.max_keys,
        is_truncated: raw.is_truncated,
        next_marker: raw.next_marker.filter(|s| !s.is_empty()),
        objects,
    })
}

#[derive(Debug, Deserialize)]
struct RawError {
    #[serde(rename = "Code")]
    code: String,
    #[serde(rename = "Message")]
    message: String,
    #[serde(rename = "RequestId", default)]
    request_id: String,
    #[serde(rename = "HostId", default)]
    host_id: String,
}

pub fn parse_error_response(xml: &str) -> Result<ErrorResponse, quick_xml::DeError> {
    let raw: RawError = from_str(xml)?;
    Ok(ErrorResponse {
        code: raw.code,
        message: raw.message,
        request_id: raw.request_id,
        host_id: raw.host_id,
    })
}

#[derive(Debug, Deserialize)]
struct RawInitiateMultipart {
    #[serde(rename = "Bucket")]
    bucket: String,
    #[serde(rename = "Key")]
    key: String,
    #[serde(rename = "UploadId")]
    upload_id: String,
}

pub fn parse_initiate_multipart_result(
    xml: &str,
) -> Result<InitiateMultipartResult, quick_xml::DeError> {
    let raw: RawInitiateMultipart = from_str(xml)?;
    Ok(InitiateMultipartResult {
        bucket: raw.bucket,
        key: raw.key,
        upload_id: raw.upload_id,
    })
}

/// Serialize a CompleteMultipartUpload body. Hand-written rather than going via
/// serde to guarantee byte-exact output (OBS rejects extra whitespace).
pub fn serialize_complete_multipart(parts: &[(u32, String)]) -> String {
    let mut out = String::from("<CompleteMultipartUpload>");
    for (n, etag) in parts {
        out.push_str(&format!(
            "<Part><PartNumber>{n}</PartNumber><ETag>{etag}</ETag></Part>"
        ));
    }
    out.push_str("</CompleteMultipartUpload>");
    out
}
